### PR TITLE
CBG-571: Code coverage and obsolete code cleanup [crud.go]

### PR DESCRIPTION
**Pull request by sarathkumarsivan**
- Removed AuthorizeDocID; obsolete.
- Removed ComputeVbSequenceChannelsForPrincipal; obsolete.
- Removed ComputeVbSequenceRolesForUser; obsolete.
- Additional coverage for deleted and listRevisions handling for get1xRevFromDoc.
- Additional coverage for the use of parentAttachmentsStruct in getAvailableRevAttachments.
- Additional coverage for checkForUpgrade.

### CODE REVIEW

File: db/crud_test.go [hunk 1]
```diff
@@ -956,3 +956,211 @@ func BenchmarkHandleRevDelta(b *testing.B) {
...
+	assert.NotEmpty(t, bodyBytes, "Document body bytes should be returned")
+	assert.False(t, removed, "This shouldn't be a removed document")
+	var response = Body{}
+	assert.NoError(t, response.Unmarshal(bodyBytes))
+	assert.Equal(t, "camera", response[BodyId])
+	assert.Equal(t, "2-a", response[BodyRev])
+	assert.Equal(t, "1483C002SKU", response["model"])
+	assert.Equal(t, "Canon EOS 5D Mark IV", response["name"])
+	assert.Equal(t, "$2,499.99", response["price"])
+	assert.Equal(t, "5578523", response["sku"])
```
Comment (adamcfraser) [inline db/crud_test.go]: Can reduce the number of properties in the doc (or at least reduce the number of properties you're asserting) - only really need one or two to verify that they are different between rev1 and rev2

File: db/crud_test.go [hunk 2]
```diff
@@ -956,3 +956,211 @@ func BenchmarkHandleRevDelta(b *testing.B) {
...
+	assert.Equal(t, "3a", response["version"])
+
+	// Create the fourth revision by deleting the third revision
+	rev4, err := db.DeleteDoc("camera", rev3)
+	assert.NoError(t, err, "Document should be deleted")
+	assert.NotEmpty(t, rev4, "Document revision shouldn't be empty")
+
+	// Get the document body bytes with the deleted revision, list revisions enabled.
+	bodyBytes, removed, err = db.get1xRevFromDoc(doc3, rev3, true)
```
Comment (adamcfraser) [inline db/crud_test.go]: rev3 shouldn't be a tombstone, should it?  I'd expect rev4 to be the tombstone.

File: db/crud_test.go [hunk 3]
```diff
@@ -956,3 +956,211 @@ func BenchmarkHandleRevDelta(b *testing.B) {
...
+	assert.NoError(t, err, "Couldn't create database 'db'")
+
+	// Create the first revision of the document
+	payload := `{"model":"1483C002SKU","sku":"5578523","name":"Canon EOS 5D Mark IV","price":"$2,499.99","version":"1a"}`
+	doc, rev1, err := db.PutExistingRevWithBody("camera", unjson(payload), []string{"1-a"}, false)
+	assert.NoError(t, err, "Couldn't create document")
+	assert.NotEmpty(t, doc, "Document shouldn't be empty")
+	assert.Equal(t, "1-a", rev1, "Provided input revision ID should be returned")
+
+	doc, buc := db.checkForUpgrade("camera", DocUnmarshalAll)
```
Comment (adamcfraser) [inline db/crud_test.go]: This doesn't look like it handles all upgrade cases - are there already other tests for that?  

This also doesn't look like it handles the differences in behaviour when running with xattrs enabled/disabled.  When enabled, the PutExistingRevWithBody would write a doc w/ xattrs, and then not need to be upgraded.
When disabled, the checkForUpgrade is going to no-op.

Review other upgrade tests for examples of how to proceed.

File: db/crud_test.go [hunk 4]
```diff
@@ -956,3 +956,211 @@ func BenchmarkHandleRevDelta(b *testing.B) {
...
+	assert.Equal(t, "camera", response[BodyId])
+	assert.Equal(t, "3-a", response[BodyRev])
+	assert.Equal(t, "1483C002SKU", response["model"])
+	assert.Equal(t, "Canon EOS 5D Mark IV", response["name"])
+	assert.Equal(t, "$2,499.99", response["price"])
+	assert.Equal(t, "5578523", response["sku"])
+	assert.Equal(t, "3a", response["version"])
+}
+
+func TestCheckForUpgrade(t *testing.T) {
```
Comment (adamcfraser) [inline db/crud_test.go]: As discussed, it looks like there should be two variations of this test:
1. Run with xattrs enabled, validate the early return from checkForUpgrade
2. Run with xattrs disabled, validate the rest of checkForUpgrade handling.

File: db/crud_test.go (hunk 2 above)
Comment (adamcfraser) [inline db/crud_test.go]: This looks like it should be targeted rev4 (the tombstone revision), and not rev 3, if it's trying to validate how get1xRevFromDoc behaves for deleted revisions.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: db/crud_test.go (hunk 1 above)
Comment (sarathkumarsivan) [reply db/crud_test.go]: Sure Adam, I have committed required changes in the latest revision.

File: db/crud_test.go (hunk 2 above)
Comment (sarathkumarsivan) [reply db/crud_test.go]: Yeah, you are right, changes are committed.

Comment (sarathkumarsivan) [reply db/crud_test.go]: This has been resolved in the latest commit.

File: db/crud_test.go (hunk 3 above)
Comment (sarathkumarsivan) [reply db/crud_test.go]: Yeah, changes were not included to handles all upgrade cases initially; but it has been added with the latest committed changes.

File: db/crud_test.go (hunk 4 above)
Comment (sarathkumarsivan) [reply db/crud_test.go]: This has been implemented and ready for review.

File: db/crud_test.go [hunk 5]
```diff
@@ -1,14 +1,14 @@
...
 import (
-	"log"
-	"testing"
-
 	"github.com/couchbase/sync_gateway/base"
 	"github.com/couchbase/sync_gateway/channels"
 	goassert "github.com/couchbaselabs/go.assert"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"log"
```
Comment (adamcfraser) [inline db/crud_test.go]: Keep the whitespace between native imports to maintain formatting.

File: db/crud_test.go [hunk 6]
```diff
@@ -956,3 +956,219 @@ func BenchmarkHandleRevDelta(b *testing.B) {
...
+	assert.NoError(t, response.Unmarshal(bodyBytes))
+
+	// Get the 1x revision from document with list revision enabled
+	bodyBytes, removed, err = db.get1xRevFromDoc(doc1, rev1, true)
+	assert.False(t, removed)
+	assert.NoError(t, err, "It should not throw any error")
+	assert.NotNil(t, bodyBytes, "Document body bytes should be received")
+	assert.NoError(t, response.Unmarshal(bodyBytes))
+
+	// Delete the revision rev2 and get the revision from doc2
```
Comment (adamcfraser) [inline db/crud_test.go]: Comment should be 'delete the document, creating tombstone revision rev3'

File: db/crud_test.go [hunk 7]
```diff
@@ -956,3 +956,219 @@ func BenchmarkHandleRevDelta(b *testing.B) {
...
+	rev3, err := db.DeleteDoc("camera", rev2)
+	bodyBytes, removed, err = db.get1xRevFromDoc(doc2, rev3, true)
+	assert.False(t, removed)
+	assert.Error(t, err, "It should throw 404 missing error")
+	assert.Nil(t, bodyBytes, "Document body bytes should be empty")
+
+	// Get the revision details from doc2 and rev2 after rev2 is deleted
```
Comment (adamcfraser) [inline db/crud_test.go]: Update comment here as well - get1xRevFromDoc for doc2 should be returning the current revision id (in this case, the tombstone revision rev3)

File: db/crud_test.go [hunk 8]
```diff
@@ -956,3 +956,219 @@ func BenchmarkHandleRevDelta(b *testing.B) {
...
+
+	// Create the first revision of the document
+	docId := "356779a9a1696714480f57fa3fb66d4c"
+	payload := `{"city":"Los Angeles"}`
+	doc, rev1, err := db.PutExistingRevWithBody(docId, unjson(payload), []string{"1-a"}, false)
+	assert.NoError(t, err, "Couldn't create document")
+	assert.NotEmpty(t, doc, "Document shouldn't be empty")
+	assert.Equal(t, "1-a", rev1, "Provided input revision ID should be returned")
+
+	// Get the 1x revision with rev1
```
Comment (adamcfraser) [inline db/crud_test.go]:
```suggestion
	// Get rev1 using get1xRevFromDoc 
```

File: db/crud_test.go [hunk 9]
```diff
@@ -956,3 +956,219 @@ func BenchmarkHandleRevDelta(b *testing.B) {
...
+	assert.Equal(t, "Los Angeles", response["city"])
+
+	// Create the second revision of the document
+	payload = `{"city":"Hollywood"}`
+	doc, rev2, err := db.PutExistingRevWithBody(docId, unjson(payload), []string{"2-a", "1-a"}, false)
+	assert.NoError(t, err, "Couldn't create document")
+	assert.NotEmpty(t, doc, "Document shouldn't be empty")
+	assert.Equal(t, "2-a", rev2, "Provided input revision ID should be returned")
+
+	// Get the 1x revision with rev2
```
Comment (adamcfraser) [inline db/crud_test.go]:
```suggestion
	// Get rev2 with get1xRevFromDoc
```

File: db/crud_test.go [hunk 10]
```diff
@@ -956,3 +956,219 @@ func BenchmarkHandleRevDelta(b *testing.B) {
...
+	assert.NotEmpty(t, bodyBytes, "Document body bytes should be returned")
+	assert.False(t, removed, "This shouldn't be a removed document")
+	assert.NoError(t, response.Unmarshal(bodyBytes))
+	assert.Equal(t, docId, response[BodyId])
+	assert.Equal(t, "2-a", response[BodyRev])
+	assert.Equal(t, "Hollywood", response["city"])
+
+	// Get body bytes from doc with unknown revision id; it simulates the error scenario.
+	// A 404 missing error should be thrown when trying get the body bytes of the document
+	// which doesn't exists in the database. The revision "3-a" doesn't exists in database.
```
Comment (adamcfraser) [inline db/crud_test.go]:
```suggestion
	// which doesn't exists in the database. The revision "3-a" doesn't exists in the revision tree
```

File: db/crud_test.go [hunk 11]
```diff
@@ -956,3 +956,219 @@ func BenchmarkHandleRevDelta(b *testing.B) {
...
+	assert.NotEmpty(t, rev3, "Document revision shouldn't be empty")
+
+	// Get the document body bytes with the deleted revision, list revisions enabled.
+	bodyBytes, removed, err = db.get1xRevFromDoc(doc, rev3, true)
+	assert.NotEmpty(t, bodyBytes, "Document body bytes should be returned")
+	assert.False(t, removed, "This shouldn't be a removed document")
+	assert.NoError(t, response.Unmarshal(bodyBytes))
+	assert.Equal(t, docId, response[BodyId])
+	assert.Equal(t, rev3, response[BodyRev])
+	assert.Equal(t, "Hollywood", response["city"])
```
Comment (adamcfraser) [inline db/crud_test.go:1151]: Validate that the BodyRevisions property is present and correct, wherever listRevisions=true is set.

File: db/crud_test.go [hunk 12]
```diff
@@ -956,3 +956,219 @@ func BenchmarkHandleRevDelta(b *testing.B) {
...
+	// when checking document revision history, it should throw 404 deleted error.
+	bodyBytes, removed, err = db.get1xRevFromDoc(doc, "", true)
+	assert.Error(t, err, "404 deleted")
+	assert.Contains(t, err.Error(), "404 deleted")
+	assert.Empty(t, bodyBytes, "Document body bytes should be empty")
+	assert.False(t, removed, "This shouldn't be a removed document")
+	assert.Error(t, response.Unmarshal(bodyBytes), "Unexpected empty JSON input to body.Unmarshal")
+}
+
+func TestCheckForUpgrade(t *testing.T) {
```
Comment (adamcfraser) [inline db/crud_test.go]: Can remove this test - should be covered by TestCheckForUpgradeOnRead in import_test.go.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: db/crud_test.go (hunk 5 above)
Comment (sarathkumarsivan) [reply db/crud_test.go]: This has been fixed in the latest revision

File: db/crud_test.go (hunk 6 above)
Comment (sarathkumarsivan) [reply db/crud_test.go]: Comment has been modified and changes are available in latest revision.

File: db/crud_test.go (hunk 7 above)
Comment (sarathkumarsivan) [reply db/crud_test.go]: Comment has been modified and changes are available in latest revision.

File: db/crud_test.go (hunk 8 above)
Comment (sarathkumarsivan) [reply db/crud_test.go]: Comment has been modified and changes are available in latest revision.

File: db/crud_test.go (hunk 9 above)
Comment (sarathkumarsivan) [reply db/crud_test.go]: Comment has been modified and changes are available in latest revision.

File: db/crud_test.go (hunk 10 above)
Comment (sarathkumarsivan) [reply db/crud_test.go]: Comment has been modified and changes are available in latest revision.

File: db/crud_test.go (hunk 12 above)
Comment (sarathkumarsivan) [reply db/crud_test.go]: This test has been removed.

File: db/crud_test.go (hunk 11 above)
Comment (sarathkumarsivan) [reply db/crud_test.go:1151]: Changed are committed to validate that the BodyRevisions property is present and correct, wherever listRevisions=true is set.

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**